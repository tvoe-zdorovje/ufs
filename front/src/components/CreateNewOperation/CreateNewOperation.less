@import "../Shared";

.panel {
  :global {
    .ant-collapse-item {
      background: #ffffff;
    }
    .ant-collapse-item-active > .ant-collapse-header {
      background: @green;
      color: #ffffff;
    }
  }
}

.rightAligned {
  padding-right: 10px;
  text-align: right;
  :global .ant-row {
    text-align: right;
  }
}

.subHead {
  height: 36px;
}

.transactionCard {
  :global .ant-row {

  }
}

.label {
  padding-left: 10px;
  vertical-align: middle;
  line-height: 32px;
  display: inline-block;
  overflow: hidden;
  white-space: nowrap;
}

.fullwidth {
  width: 100%;
  text-decoration: underline;
}

.bordered {
  border: 1px solid #e9e9e9;
  text-align: center;
}

.sticked {
  margin-bottom: 0;
}

.disabledInputRow {
  margin-bottom: 5px;
}
