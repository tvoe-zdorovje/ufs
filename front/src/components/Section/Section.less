@import "../Shared";

.section {
  margin-top: 12px;
  border: 1px solid @green;
}

.header {
  padding: 1px 8px 1px 4px;
  margin: -12px 0 0 8px;
  color: @green;
  background-color: #fff;
  position: absolute;
  z-index: 10;
}

.container {
  padding: 10px 5px 5px 5px;
}
