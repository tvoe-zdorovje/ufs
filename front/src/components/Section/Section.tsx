import React from 'react'

const styles = require('./Section.less')

interface SectionProps {
  header: string,
  children?: any,
}

export default (props: SectionProps) => (
  <section className={styles.section}>
    <div className={styles.header}>{props.header}</div>
    <div className={styles.container}>
      {props.children}
    </div>
  </section>
);
