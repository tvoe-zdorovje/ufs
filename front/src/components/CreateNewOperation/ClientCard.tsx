import React from 'react'
import {
  Col,
  Input,
  Row,
  Checkbox
} from 'antd'

import Section from 'components/Section/Section'
import { Representative } from 'interfaces/dto/Representative.interface'

const styles = require('./CreateNewOperation.less')

export default (props: Representative) => {
  const disabledInput = value => (
    <Input disabled value={value} />
  )

  const label = (value) => (<span className={styles.label}>{value}</span>)

  return (
    <div style={{ minWidth: 300 }}>
      <Section header="">
        <Row>
          <Col span={7}>
            {label('Фамилия')}
          </Col>
          <Col span={16}>
            {disabledInput(props.lastName)}
          </Col>
        </Row>
        <Row>
          <Col span={7}>
            {label('Имя')}
          </Col>
          <Col span={16}>
            {disabledInput(props.firstName)}
          </Col>
        </Row>
        <Row>
          <Col span={7}>
            {label('Отчество')}
          </Col>
          <Col span={16}>
            {disabledInput(props.patronymic)}
          </Col>
        </Row>
        <Row>
          <Col span={7}>
            {label('Дата рождения')}
          </Col>
          <Col span={16}>
            {disabledInput(props.birthDate)}
          </Col>
        </Row>
        <Row>
          <Col span={7}>
            {label('Тип ДУЛ')}
          </Col>
          <Col span={16}>
            {disabledInput(props.document.type)}
          </Col>
        </Row>
        <Row>
          <Col span={7}>
            {label('Серия')}
          </Col>
          <Col span={7}>
            {disabledInput(props.document.series)}
          </Col>
          <Col span={4}>
            {label('Номер')}
          </Col>
          <Col span={5}>
            {disabledInput(props.document.number)}
          </Col>
        </Row>
        <Row>
          <Col span={7}>
            {label('Место рождения')}
          </Col>
          <Col span={16}>
            {disabledInput(props.birthPlace)}
          </Col>
        </Row>
        <Row>
          <Col span={7}>
            {label('Дата выдачи')}
          </Col>
          <Col span={16}>
            {disabledInput(props.document.issuedDate)}
          </Col>
        </Row>
        <Row>
          <Col span={7}>
            {label('Выдан (орган)')}
          </Col>
          <Col span={16}>
            {disabledInput(props.document.issuedBy)}
          </Col>
        </Row>
        <Row>
          <Col span={7}>
            {label('ИНН/КИО')}
          </Col>
          <Col span={16}>
            {disabledInput(props.inn)}
          </Col>
        </Row>
        <Row>
          <Col span={7}>
            {label('Индекс')}
          </Col>
          <Col span={16}>
            {disabledInput(props.postcode)}
          </Col>
        </Row>
        <Row>
          <Col span={7}>
            {label('Адрес')}
          </Col>
          <Col span={16}>
            {disabledInput(props.address)}
          </Col>
        </Row>
        <Row>
          <Col span={7}>
            {label('Резидент')}
          </Col>
          <Col span={16}>
            <Checkbox checked={props.resident} />
          </Col>
        </Row>
      </Section>
    </div>
  )
}
