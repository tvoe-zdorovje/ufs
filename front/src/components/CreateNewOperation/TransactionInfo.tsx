import React from 'react'
import {
  Form,
  Col,
  Input,
  Row,
  Button,
  Modal,
} from 'antd'
const FormItem = Form.Item;

import Section from 'components/Section/Section'
import ClientCard from  './ClientCard'
import { Task } from 'interfaces/dto/Task.interface'
import { Representative } from 'interfaces/dto/Representative.interface'
import m2w from 'utils/MoneyToWords'

const styles = require('./CreateNewOperation.less')

export interface TransactionInfoProps {
  getFieldDecorator: Function,
  representative: Representative,
  task: Task,
  ovnBased?: boolean,
  readOnly?: boolean,
}

export default class TransactionInfo extends React.Component<TransactionInfoProps, any> {
  state = { visible: false }
  toggleModal = () => {
    this.setState({
      visible: !this.state.visible,
    });
  }
  handleCancel = (e) => {
    this.setState({
      visible: false,
    });
  }

  render() {
    const { getFieldDecorator, readOnly = false, ovnBased = false, representative } = this.props
    const task = { ...this.props.task, amountInWords: m2w(this.props.task.amount) }
    const disabledInput = (key) => (<FormItem>
      {(typeof task[key] === 'undefined') ?
        <Input disabled />
        :
        <Input disabled value={task[key]} />}
    </FormItem>)

    const requiredInput = (key, emptyMessage) => {
      if (readOnly) {
        return disabledInput(key)
      }
      let config = { rules: [{ required: true, message: emptyMessage }] }
      if (typeof task[key] !== 'undefined') {
        config = { ...config, initialValue: task[key] }
      }
      return (
        <FormItem>
          {getFieldDecorator(key, config)(
            <Input />
          )}
        </FormItem>
      )
    }

    const amountInput = () => {
      const key = 'amount'
      if (readOnly || ovnBased) {
        return (<Input disabled value={`${task.amount} 00 коп.`} />)
      }
      let config = {
        rules: [{ required: true, message: 'Укажите сумму операции' },
          { pattern: (new RegExp('^\\d+$')), message: 'сумма операции должна быть целым числом' }],
      }
      if (typeof task.amount !== 'undefined') {
        config = { ...config, initialValue: task.amount }
      }
      return (
        <FormItem>
          {getFieldDecorator(key, config)(
            <Input addonAfter=" 00 коп." />
          )}
        </FormItem>
      )
    }

    const label = (value) => (<span className={styles.label}>{value}</span>)
    return (
      <div className="transactionCard">
        <Section header="Основная информация">
          <Row>
            <Col span={24}>
              <Row>
                <Col span={10}>
                  {label('Объявление №')}
                  {ovnBased ?
                    label(task.num)
                    :
                    requiredInput('num', 'Укажите номер объявления')
                  }
                </Col>
                <Col span={14}>
                  {label(`Дата создания ${task.createdDate}`)}
                </Col>
              </Row>
              <Row>
                <Col span={3}>
                  {label('От кого')}
                </Col>
                <Col span={13}>
                  <Button type="dashed"
                          className={styles.fullwidth}
                          onClick={this.toggleModal}
                  >{task.repFio}</Button>
                  <Modal
                    title="Карточка клиента/представителя клиента"
                    visible={this.state.visible}
                    onCancel={this.handleCancel}
                    cancelText="Отмена"
                    footer={null}
                  >
                    <ClientCard {...representative} />
                  </Modal>
                </Col>
                <Col span={2} className={styles.rightAligned}>
                  <div className={styles.subHead}></div>
                  <div>{label('счет №')}</div>
                </Col>
                <Col span={6}>
                  <div className={styles.subHead}>Дебет</div>
                  {disabledInput('debetAccount')}
                </Col>
              </Row>
              <Row>
                <Col span={3}>
                  {label('Получатель')}
                </Col>
                <Col span={13}>
                  {disabledInput('legalEntityShortName')}
                </Col>
                <Col span={2} className={styles.rightAligned}>
                  <div className={styles.subHead}></div>
                  <div>{label('счет №')}</div>
                </Col>
                <Col span={6}>
                  <div className={styles.subHead}>Кредит</div>
                  {disabledInput('accountId')}
                </Col>
              </Row>
              <Row>
                <Col span={1}>
                  {label('ИНН')}
                </Col>
                <Col span={4}>
                  {disabledInput('inn')}
                </Col>
                <Col span={1}>
                  {label('КПП')}
                </Col>
                <Col span={4}>
                  {disabledInput('kpp')}
                </Col>
                <Col span={2} className={styles.rightAligned}>
                  {label('счет №')}
                </Col>
                <Col span={6}>
                  {disabledInput('accountId')}
                </Col>
                <Col span={6} />
              </Row>
              <Row>
                <Col span={6}>
                  {label('Наименование банка-вносителя')}
                </Col>
                <Col span={12}>
                  {disabledInput('senderBank')}
                </Col>
                <Col span={2} className={styles.rightAligned}>
                  {label('БИК')}
                </Col>
                <Col span={4}>
                  {disabledInput('senderBankBic')}
                </Col>
              </Row>
              <Row>
                <Col span={6}>
                  {label('Наименование банка-получателя')}
                </Col>
                <Col span={12}>
                  {disabledInput('recipientBank')}
                </Col>
                <Col span={2} className={styles.rightAligned}>
                  {label('БИК')}
                </Col>
                <Col span={4}>
                  {disabledInput('recipientBankBic')}
                </Col>
              </Row>
              <Row>
                <Col span={4}>
                  {label('Сумма операции')}
                </Col>
                <Col span={6}>
                  {amountInput()}
                </Col>
              </Row>
              <Row>
                <Col span={4}>
                  {label('Сумма прописью')}
                </Col>
                <Col span={20}>
                  {disabledInput('amountInWords')}
                </Col>
              </Row>
              <Row>
                {label('Источник поступления')}
              </Row>
              <Row>
                {ovnBased ?
                  disabledInput('source')
                  :
                  requiredInput('source', 'Укажите источник поступения')
                }
              </Row>
            </Col>
          </Row>
        </Section>
        { task.clientTypeFk ?
        <Section header="Дополнительная информация">
          <Row>
            {label('Наименование организации')}
          </Row>
          <Row>
            {requiredInput('organisationNameFk', 'Укажите наименование организации')}
          </Row>
          <Row>
            {label('Лицевой счет')}
          </Row>
          <Row>
            {requiredInput('personalAccountId', 'Укажите лицевой счет')}
          </Row>
        </Section>
          : null
        }
      </div>
    )
  }
}
