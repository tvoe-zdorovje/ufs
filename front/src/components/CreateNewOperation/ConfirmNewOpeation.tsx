import React from 'react'
import {
  Collapse,
  Col,
  Form,
  Row,
  Button
} from 'antd'

import TransactionInfo from './TransactionInfo'
import cancelOperation from '../CancelOperation'
import { CreateNewOperationFormProps } from './CreateNewOperation'
import { Representative } from 'interfaces/dto/Representative.interface'

const styles = require('./CreateNewOperation.less')

export interface ConfirmNewOperationProps extends CreateNewOperationFormProps {
  onComplete: Function,
  onCancel: Function,
  changeTransaction: any,
  representative: Representative,
  commission: string,
  packageId?: string,
  operationTypeCode?: string,
  form?: any,
}

@Form.create<any>()
export default class ConfirmNewOperation extends React.Component<ConfirmNewOperationProps, any> {
  handleSubmit = (e) => {
    e.preventDefault()
    this.props.form.validateFields((err) => {
      if (!err) {
        this.props.onComplete()
      }
    })
  }

  render() {
    const { getFieldDecorator } = this.props.form
    const { task, representative, onCancel, commission, changeTransaction } = this.props
    return (
      <div>
        <Form onSubmit={this.handleSubmit}>
          <Row>
            <Col span={18}>
              <Collapse className={styles.panel} defaultActiveKey={['1']}>
                <Collapse.Panel header="Реквизиты операции" key="1">
                  <TransactionInfo
                    getFieldDecorator={getFieldDecorator}
                    task={task}
                    representative={representative}
                    readOnly
                  />
                </Collapse.Panel>
              </Collapse>
              <Row className={styles.rightAligned}>
                <Button htmlType="submit" className={styles.greenButton}>
                  Перейти к подтверждению
                </Button>
                <Button
                  className={styles.orangeButton}
                  onClick={changeTransaction}
                >
                  Изменить
                </Button>
                <Button className={styles.redButton} onClick={() => cancelOperation(onCancel)}>
                  Отменить операцию
                </Button>
              </Row>
            </Col>
            <Col span={6}>
              <Collapse className={styles.panel} defaultActiveKey={['1']}>
                <Collapse.Panel header="Комиссия" key="1">
                  <div>Размер комиссии</div>
                  <div>{`${commission} рублей 00 коп.`}</div>
                </Collapse.Panel>
              </Collapse>
            </Col>
          </Row>
        </Form>
      </div>
    )
  }
}
