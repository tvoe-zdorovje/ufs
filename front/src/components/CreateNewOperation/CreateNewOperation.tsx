import React from 'react'
import {
  Button,
  Collapse,
  Col,
  Form,
  Modal,
  Radio,
  Row
} from 'antd'

import TransactionInfo from './TransactionInfo'
import cancelOperation from 'components/CancelOperation'
import { Task } from 'interfaces/dto/Task.interface'
import { Representative } from 'interfaces/dto/Representative.interface'
import { Subbranch } from 'interfaces/dto/Subbranch.interface'
import { OVN } from 'interfaces/dto/OVN.interface'
import { loadCommission } from '../../reducers/shared/createNewOperation'

const styles = require('./CreateNewOperation.less')

export interface CreateNewOperationFormProps {
  continueUrm: Function,
  continueCache: Function,
  onCancel: Function,
  clearTransaction: Function,
  onAlter: Function,
  loadCommission: Function,
  loadRepresentativeAndTask: Function,
  ovnBased: boolean,
  taskLoaded: boolean
  commissionLoading: boolean,
  subbranch: Subbranch,
  ovn?: OVN,
  representative: Representative,
  limit: string,
  task?: Task
  workplaceId?: string,
  form?: any,
}

const checkLimits = (amount, limit) => {
  try {
    const amountInt = parseInt(amount, 10)
    const limitInt = parseInt(limit, 10)
    return amountInt <= limitInt
  } catch ( e ) {
    return false
  }
}

@Form.create<any>()
export default class CreateNewOperation extends React.Component<CreateNewOperationFormProps, any> {
  constructor(props) {
    super(props)
    this.onOk = this.onOk.bind(this)
  }

  state = {
    confirmType: 0,
  }

  componentWillMount() {
    const { subbranch, ovn, taskLoaded, loadRepresentativeAndTask } = this.props
    if (!taskLoaded) {
      loadRepresentativeAndTask(subbranch, ovn)
    }
  }

  onChange = (e) => {
    this.setState({
      confirmType: e.target.value,
    })
  }

  onOk = (values) => {
    const { workplaceId, onAlter, continueUrm, continueCache } = this.props
    const isUrm = this.state && this.state.confirmType === 1
    if (isUrm) {
      continueUrm(values, workplaceId)
    } else {
      continueCache(values, workplaceId, onAlter)
    }
  }

  confirmDialogContent = (isLimited) => (
    <Radio.Group onChange={this.onChange}>
      { isLimited ? <Radio value={1}>Продолжить операцию на данном УРМ</Radio> : null }
      <Radio value={2}>Перенаправить клиента в кассу для продолжения операции</Radio>
    </Radio.Group>
  )

  confirm = () => {
    const { form, limit } = this.props
    const values = form.getFieldsValue()
    form.validateFields((err) => {
      if (!err) {
        this.props.loadCommission(values, this.handleSubmit)
      }
    })
  }

  handleSubmit = (values: any, isUrm: boolean) => {
    const { form, task, limit } = this.props
    const onOk = this.onOk
    const isLimited = checkLimits(values.amount || task.amount, limit)
    Modal.confirm({
      title: 'Выбор рабочего места продолжения операции',
      content: this.confirmDialogContent(isLimited && isUrm),
      okText: 'ОК',
      cancelText: 'Отмена',
      width: 600,
      onOk() {
        onOk(values)
      },
    })

  }

  handleClear = () => {
    const { form, clearTransaction } = this.props
    form.resetFields()
    clearTransaction()
  }

  render() {
    const { getFieldDecorator } = this.props.form
    const { ovnBased, onCancel, representative, task, commissionLoading } = this.props
    return (
      <div>
        <Form>
          <Row>
            <Col span={18}>
              <Collapse className={styles.panel} defaultActiveKey={['1']}>
                <Collapse.Panel header="Реквизиты операции" key="1">
                  {task ?
                    <TransactionInfo
                      getFieldDecorator={getFieldDecorator}
                      task={task}
                      representative={representative}
                      ovnBased={ovnBased}
                    />
                    :
                    null
                  }
                </Collapse.Panel>
              </Collapse>
              {this.props.taskLoaded ? <Row className={styles.rightAligned}>
                <Button loading={commissionLoading} onClick={this.confirm} className={styles.greenButton}>Подтвердить</Button>
                <Button
                  className={styles.orangeButton}
                  onClick={this.handleClear}
                >
                  Очистить форму
                </Button>
                <Button
                  className={styles.redButton}
                  onClick={() => cancelOperation(onCancel)}
                >
                  Отменить операцию
                </Button>
              </Row> : <Row />
              }
            </Col>
            <Col span={6} />
          </Row>
        </Form>
      </div>
    )
  }
}

