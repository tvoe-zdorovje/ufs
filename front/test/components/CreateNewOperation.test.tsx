import React from 'react'
import {
  render,
  shallow,
  mount,
} from 'enzyme'
import toJson from 'enzyme-to-json'

import CreateNewOperation, { CreateNewOperationFormProps } from 'components/CreateNewOperation/CreateNewOperation'
import Section from 'components/Section/Section'

const mockFn = () => {
}
describe('Компонент CreateNewOperation', () => {
  const taskValues = {
    accountInfo: {
      id: '4782 9871 9873 897009',
      typeId: 'Расчетный счет',
      agreement: {
        id: '124567',
        openDate: '',
      },
      status: '1',
      lastTransactionDate: '18.06.2001',
      currencyType: 'RUB',
      active: true,
      tbCode: '',
      gosbCode: '',
      vspCode: '',
      subbranchCode: '',
      changeStatusDescription: '',
      clientTypeFk: true,
      seizured: true,
    },
    subbranch: {
      bankName: '',
      bic: '',
      gosbCode: '',
      inn: '',
      level: '',
      subbranchCode: '',
      tbCode: '',
      vspCode: '',
    },
  }

  const props: CreateNewOperationFormProps = {
    continueUrm: mockFn,
    continueCache: mockFn,
    onCancel: mockFn,
    onAlter: mockFn,
    clearTransaction: mockFn,
    loadRepresentativeAndTask: mockFn,
    limit: '100',
    taskLoaded: false,
    ovnBased: false,
    subbranch: taskValues.subbranch,
    representative: {
      inn: '',
      lastName: 'ln',
      firstName: '',
      birthDate: '',
      birthPlace: '',
      resident: true,
      document: {
        type: '',
        series: '',
        number: '',
        issuedBy: '',
        issuedDate: '',
      },
    },
  }

  const task = {
    senderBank: '',
    senderBankBic: '',
    recipientBank: '',
    recipientBankBic: '',
    currencyType: '',
    num: '1',
    accountId: '',
    legalEntityShortName: '',
    inn: '',
    createdDate: '18.05.2017 18:00:00',
    repFio: `1 2 3`,
    symbols: [],
  }
  const wrapper = shallow(<CreateNewOperation {...props} />)
  const renderWrapper = render(<CreateNewOperation {...props} />)

  it('отображает компонент Form', () => {
    const html = wrapper.html()
    expect(html.match(/class="ant-form\sant-form-horizontal"/g) || [])
      .toHaveLength(1)
  })

  it('не отображает компонент <TransactInfo /> при отсутствии задачи', () => {
    const html = wrapper.html()
    expect(html.match(/class="transactionCard"/g) || [])
      .toHaveLength(0)
  })

  it('отображает 3 кнопки если задача загружена ', () => {
    wrapper.setProps({ taskLoaded: true })
    wrapper.update()
    const html = wrapper.html()
    expect(html.match(/class="ant-btn/g) || [])
      .toHaveLength(3)
  })

  it('отображает компонент <TransactInfo /> при наличии задачи>', () => {
    wrapper.setProps({ task, taskLoaded: true })

    wrapper.update()
    const html = wrapper.html()
    expect(html.match(/class="transactionCard"/g) || [])
      .toHaveLength(1)
  })

  it('отображает компоненты Section', () => {
    expect(wrapper.find(Section))
      .toHaveLength(0)
  })

  it('запрашивает loadRepresentativeAndTask если не загружено', () => {
    const spy = jest.fn()
    const getCreateNewOperationSpy = { ...props, loadRepresentativeAndTask: spy }
    mount(<CreateNewOperation {...getCreateNewOperationSpy} />)
    expect(spy).toBeCalled()
  })

  it('совпадает со своим снимком', () => {
    expect(toJson(renderWrapper)).toMatchSnapshot()
  })
})
