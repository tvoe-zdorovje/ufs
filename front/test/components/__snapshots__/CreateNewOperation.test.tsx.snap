// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Компонент CreateNewOperation совпадает со своим снимком 1`] = `
<div>
  <form
    class="ant-form ant-form-horizontal"
  >
    <div
      class="ant-row"
    >
      <div
        class="ant-col-18"
      >
        <div
          class="ant-collapse panel"
        >
          <div
            class="ant-collapse-item ant-collapse-item-active"
          >
            <div
              aria-expanded="true"
              class="ant-collapse-header"
              role="tab"
            >
              <i
                class="arrow"
              />
              Реквизиты операции
            </div>
            <div
              class="ant-collapse-content ant-collapse-content-active"
              role="tabpanel"
            >
              <div
                class="ant-collapse-content-box"
              />
            </div>
          </div>
        </div>
        <div
          class="ant-row"
        />
      </div>
      <div
        class="ant-col-6"
      />
    </div>
  </form>
</div>
`;
